File: explain_this_repo/providers/ollama.py
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Dict

from explain_this_repo.providers.base import LLMProvider, LLMProviderError

DEFAULT_MODEL = "llama3"
DEFAULT_HOST = "http://localhost:11434"


class OllamaProvider(LLMProvider):
    name = "ollama"

    def __init__(self, config: Dict[str, Any]) -> None:
        self.model = config.get("model", DEFAULT_MODEL)
        self.host = config.get("host", DEFAULT_HOST).rstrip("/")

        self.validate_config()

    def validate_config(self) -> None:
        if not self.host.startswith("http"):
            raise LLMProviderError(
                "Ollama host must be a valid URL (e.g. http://localhost:11434)"
            )

    def doctor(self) -> list[str]:
        results = []

        url = f"{self.host}/api/tags"

        try:
            req = urllib.request.Request(url, method="GET")
            with urllib.request.urlopen(req, timeout=5) as response:
                if response.status == 200:
                    results.append("Ollama server reachable")
                else:
                    results.append(f"Ollama server responded with {response.status}")
        except Exception:
            results.append("Ollama server not reachable")

        results.append(f"model: {self.model}")
        results.append(f"host: {self.host}")

        return results

    def generate(self, prompt: str) -> str:
        url = f"{self.host}/api/generate"

        payload = {
            "model": self.model,
            "prompt": prompt,
            "stream": False,
        }

        req = urllib.request.Request(
            url,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"},
            method="POST",
        )

        try:
            with urllib.request.urlopen(req, timeout=120) as response:
                raw = response.read().decode("utf-8")
        except urllib.error.URLError as e:
            raise LLMProviderError(
                "Failed to connect to Ollama.\n"
                "Ensure Ollama is running locally.\n"
                "Start it with: ollama serve"
            ) from e
        except Exception as e:
            raise LLMProviderError(f"Ollama request failed: {e}") from e

        try:
            data = json.loads(raw)
        except Exception as e:
            raise LLMProviderError("Invalid response from Ollama") from e

        text = data.get("response")
        if not text or not text.strip():
            raise LLMProviderError("Ollama returned no text")

        return text.strip()

File: node_version/providers/ollama.ts
import { LLMProvider, LLMProviderError } from "./base.js"

const DEFAULT_HOST = "http://localhost:11434"

type OllamaConfig = {
  model?: string
  host?: string
}

export class OllamaProvider implements LLMProvider {

  name = "ollama"

  private model: string
  private host: string

  constructor(config: OllamaConfig = {}) {

    this.model = config.model as string
    this.host = (config.host ?? DEFAULT_HOST).replace(/\/$/, "")

    this.validateConfig()
  }

  validateConfig(): void {

    if (!this.model || !String(this.model).trim()) {
      throw new LLMProviderError(
        "Ollama provider requires a model.\n" +
        "Set providers.ollama.model (e.g. llama3, gemma3:4b, glm-5:cloud)."
      )
    }

    if (!this.host.startsWith("http")) {
      throw new LLMProviderError(
        "Ollama host must be a valid URL (e.g. http://localhost:11434)"
      )
    }
  }

  async doctor(): Promise<string[]> {

    const results: string[] = []

    try {

      const res = await fetch(`${this.host}/api/tags`, {
        method: "GET"
      })

      if (res.ok) {
        results.push("Ollama server reachable")
      } else {
        results.push(`Ollama server responded with ${res.status}`)
      }

    } catch (err: any) {
      results.push(`Ollama server not reachable: ${String(err?.message ?? err)}`)
    }

    results.push(`model: ${this.model}`)
    results.push(`host: ${this.host}`)

    return results
  }

  async generate(prompt: string): Promise<string> {

    const url = `${this.host}/api/generate`

    const payload = {
      model: this.model,
      prompt,
      stream: false
    }

    let res: Response

    try {
      res = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(payload)
      })
    } catch (err: any) {
      throw new LLMProviderError(
        [
          "Failed to connect to Ollama.",
          "Ensure Ollama is running locally.",
          "Start it with: ollama serve",
          `Error: ${String(err?.message ?? err)}`
        ].join("\n")
      )
    }

    if (!res.ok) {
      throw new LLMProviderError(
        `Ollama HTTP error: ${res.status}`
      )
    }

    let data: any

    try {
      data = await res.json()
    } catch {
      throw new LLMProviderError("Invalid response from Ollama")
    }

    const text = data?.response

    if (!text || !String(text).trim()) {
      throw new LLMProviderError("Ollama returned no text")
    }

    return String(text).trim()
  }
}